### PR TITLE
[Enhance] MaskFormer refactor

### PR DESCRIPTION
**Pull request by chhluo**
## Motivation

MaskFormer refactor, part of #6938 

## Modification

1. a new panoptic fusion head `MaskFormerFusionHead` which will be used by MaskFormer and Mask2Former, containing `instance_postprocess`, `panoptic_postprocess`, `semantic_postprocess`(not implemanted yet), and move the panoptic_postprocess from `MaskFormerHead` to `MaskFormerFusionHhead`.
2. update the config.
3. update `simple_test` in MaskFormer detector, which will be reused by Mask2Former.
4. `mask2bbox` which generated bboxes from mask, may be used by other model in the future.
5. unit test

### CODE REVIEW

**Review by ZwwWayne — APPROVED** [review]

File: mmdet/models/seg_heads/panoptic_fusion_heads/maskformer_fusion_head.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+                 num_stuff_classes=53,
+                 test_cfg=None,
+                 loss_panoptic=None,
+                 init_cfg=None,
+                 **kwargs):
+        super().__init__(num_things_classes, num_stuff_classes, test_cfg,
+                         loss_panoptic, init_cfg, **kwargs)
+
+    def forward_train(self, **kwargs):
+        """HeuristicFusionHead has no training loss."""
```
Comment (AronLin) [inline mmdet/models/seg_heads/panoptic_fusion_heads/maskformer_fusion_head.py:25]: HeuristicFusionHead?

File: configs/maskformer/maskformer_r50_mstrain_16x1_75e_coco.py [hunk 2]
```diff
@@ -116,8 +123,19 @@
...
+        # For now, the dataset does not support
+        # evaluating semantic segmentation metric.
+        semantic_on=False,
+        instance_on=False,
+        # max_per_image is for instance segmentation.
+        max_per_image=100,
+        object_mask_thr=0.8,
+        iou_thr=0.8,
+        # In MaskFormer's panoptic postprocessing,
+        # it will not filter mask area where score is less than 0.5 .
```
Comment (ZwwWayne) [inline configs/maskformer/maskformer_r50_mstrain_16x1_75e_coco.py:137]: it will not filter masks whose score is smaller than 0.5

File: mmdet/core/mask/utils.py [hunk 3]
```diff
@@ -62,3 +63,27 @@ def encode_mask_results(mask_results):
         return encoded_mask_results, cls_mask_scores
     else:
         return encoded_mask_results
+
+
+def mask2bbox(masks):
+    """Tight bounding boxes around mask.
```
Comment (ZwwWayne) [inline mmdet/core/mask/utils.py:69]: Obtain tight bounding boxes of binary masks.

File: tests/test_utils/test_masks.py [hunk 4]
```diff
@@ -687,3 +687,14 @@ def test_polygon_mask_iter():
...
     for i, polygon_mask in enumerate(polygon_masks):
         assert np.equal(polygon_mask, raw_masks[i]).all()
+
+
+def test_mask2bbox():
+    masks = torch.zeros((1, 20, 15), dtype=torch.bool)
+    # [x1, y1, x2, y2]
+    bboxes_gt = torch.tensor([[10, 10, 15, 20]])
+    masks[0, bboxes_gt[0, 1]:bboxes_gt[0, 3],
+          bboxes_gt[0, 0]:bboxes_gt[0, 2]] = True
```
Comment (AronLin) [inline tests/test_utils/test_masks.py:697]: Perhaps a sample of arbitrary shapes is better.

**Review by AronLin — COMMENTED** [review]
Did you evaluate again, did the code refactoring affect the results?

**Comment by chhluo** [comment]
> Did you evaluate again, did the code refactoring affect the results?

Yes, same as original result.

**Review by AronLin — APPROVED** [review]
LGTM

**Comment by AronLin** [comment]
Kindly ping @ZwwWayne to final check.

**Comment by chhluo** [comment]
> Kindly ping @ZwwWayne to final check.

After refactor, the config file in readme should be updated.

File: mmdet/models/dense_heads/maskformer_head.py [hunk 5]
```diff
@@ -116,12 +116,8 @@ def __init__(self,
         self.test_cfg = test_cfg
         self.train_cfg = train_cfg
         if train_cfg:
-            assert 'assigner' in train_cfg, 'assigner should be provided '\
-                'when train_cfg is set.'
-            assigner = train_cfg['assigner']
-            self.assigner = build_assigner(assigner)
-            sampler_cfg = dict(type='MaskPseudoSampler')
```
Comment (chhluo) [inline mmdet/models/dense_heads/maskformer_head.py:125]: remove redundant code

**Review by ZwwWayne — APPROVED** [review]
